---
title: "Supplementary material for Biology Background"
author: "Peter Hickey"
date: "Modified: 22 April 2015. Compiled: `r format(Sys.Date(), '%d %b %Y')`"
output:
  html_document:
    keep_md: yes
---

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

# CpGs are underrepresented in the human genome (hg19)

Firstly, the frequency of each nucleotide is computed. Secondly, the expected frequency of each dinucleotide is computed under the _independence model_. Under the independence model, the frequency of each dinucleotide is simply the product of the respective observed nucleotide frequencies. Finally, these expected frequencies are compared to the observed values from the respective reference genomes.

I only use the autosomes, sex chromosomes and mitochondrial chromosome in all analyses, that is, I do not consider the "random", "unplaced" and "alternate haplotype" contigs in the reference genomes.

All results are computed at the chromosome-level as well as being summarised genome-wide.

## Compute the nucleotide and dinucleotide frequencies

```{r, cache = TRUE, message = FALSE}
library(BSgenome.Hsapiens.UCSC.hg19)
library(knitr)
# Nucleotide frequencies
nf <- bsapply(new("BSParams", X = Hsapiens, FUN = letterFrequency,
                  exclude = c("rand", "Un", "hap"), simplify = TRUE),
              letters = c("A", "C", "G", "T"))
# Pretty print the results
kable(data.frame(percentage = round(100 * rowSums(nf) /
                                      sum(as.numeric(nf)), 1)))

# GC-percentage
round(100 * sum(nf[c("C", "G"), ]) / sum(as.numeric(nf)), 1)

# Dinucleotide frequencies
df <- bsapply(new("BSParams", X = Hsapiens, FUN = dinucleotideFrequency,
                  exclude = c("rand", "Un", "hap"), simplify = TRUE))
# Pretty print the results
kable(data.frame(percentage = round(100 * rowSums(df) /
                                      sum(as.numeric(df)), 1)))

# CpG percentage
round(100 * rowSums(df) / sum(as.numeric(df)), 1)["CG"]
```

## Plot the observed:expected ratio for each dinucleotide

```{r dinucleotide_distribution, fig.lp = "fig:dinuclotide_distribution", dev = "pdf", message = FALSE}
library(ggplot2)
thesis_theme <- theme_classic(base_size = 20)
thesis_theme <- theme_bw(base_size = 20)
edf <- nf %*% t(nf) / (sum(nf %*% t(nf))) * 100
x <- data.frame(Dinucleotide = names(rowSums(df)),
                Observed = as.vector(100 * rowSums(df) / sum(as.numeric(df))),
                Expected = as.vector(edf))
g <- ggplot(x, aes(x = Dinucleotide, y = Observed / Expected)) +
  geom_point(size = 5) + coord_cartesian(ylim = c(0, 1.5)) +
  ggtitle("CpGs are underrepresented in human reference genome (hg19)") +
  thesis_theme +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  geom_hline(yintercept = 1, linetype = 2)
g
g
ggsave("dinucleotide_frequency_hg19.pdf", g, height = 9, width = 16)
```

## Frequency of CpGs on each chromosome in the human reference genome (hg19)

Chromosome 19 has the highest CpG density of the autosomes in the human reference genome (hg19):
```{r cpg_frequency}
cpg_density <- data.frame(Chromosome = colnames(df), 
                          Percentage = df['CG', ] / colSums(df) * 100)
knitr::kable(cpg_density, align = "c")
write.table(cpg_density, "CpG_density_hg19.txt", col.names = TRUE, 
            row.names = FALSE, quote = FALSE)
```

# Distances between CpGs in the human reference genome (hg19)

I extract the co-ordinates of all CpGs in the human reference genome (hg19) and compute the intra-pair distances (IPD) between adjacent CpGs. I then plot the empirical cumulative distribution function of IPD. These are compared to the distances we would expect if CpGs were uniformly distributed across the genome.

I only use the autosomes, sex chromosomes and mitochondrial chromosome in all analyses, that is, I do not consider the "random", "unplaced" and "alternate haplotype" contigs in the reference genomes.

```{r cpg_ipds, cache = FALSE, message = FALSE}
set.seed(666)

# Observed IPDs
cpgs <- bsapply(new("BSParams", X = Hsapiens, FUN = matchPattern,
                    exclude = c("rand", "Un", "hap"), simplify = TRUE),
                pattern = "CG")
d <- unlist(lapply(cpgs, start), use.names = FALSE)
ipd <- diff(d)
# Drop negative distances, which correspond to CpGs on different chromosomes
ipd <- ipd[ipd > 0]

# Expected IPDs if CpGs were uniformly distributed on genome
d_sim <- bsapply(new("BSParams", X = Hsapiens, FUN = function(x) {
  cpg_freq <- countPattern("CG", x)
  sort(sample(x = length(x), size = cpg_freq, replace = FALSE))
  }, exclude = c("rand", "Un", "hap"), simplify = TRUE))
ipd_sim <- diff(unlist(d_sim, use.names = FALSE))
# Drop negative distances, which correspond to CpGs on different chromosomes
ipd_sim <- ipd_sim[ipd_sim > 0]

# Drop negative values, which are due to CpGs being on different chromosomes
ipd_df <- data.frame(IPD = c(ipd, ipd_sim),
                     Source = c(rep("Observed", length(ipd)),
                                rep("Expected", length(ipd_sim))))
```

## Empirical cumulative distribution plot

```{r cpg_ipds_cdf, dev = "pdf", message = FALSE}
library(RColorBrewer)
thesis_theme <- theme_classic(base_size = 20)
thesis_theme <- theme_bw(base_size = 20)

g <- ggplot(aes(x = IPD, colour = Source), data = ipd_df) + 
  stat_ecdf(size = 1.5) + 
  thesis_theme +
  coord_cartesian(xlim = c(0, 1000)) +
  xlab("Distance between adjacent CpGs (bp)\n(Axis truncated at 1000)") +
  ggtitle("Human reference genome (hg19)") + 
  scale_color_brewer(palette = "Dark2") + 
  ylab("Proportion of CpGs within given distance")
g
ggsave("IPD_ECDF_hg19.pdf", g, width = 16, height = 9)
```

## Frequency polygon

```{r cpg_ipds_hist, dev = "pdf"}
library(scales)

ipd_table <- table(ipd)
ipd_sim_table <- table(ipd_sim)
ipd_table_df <- data.frame(IPD = c(as.numeric(names(ipd_table)),
                                   as.numeric(names(ipd_sim_table))),
                           Count = c(ipd_table, ipd_sim_table),
                           Source = c(rep("Observed", length(ipd_table)),
                                      rep("Expected", length(ipd_sim_table))))

g <- ggplot(aes(x = IPD, y = Count),
            data = subset(ipd_table_df, Source == "Observed")) +
  geom_line(size = 1.2) + coord_cartesian(xlim = c(0, 300)) +
  xlab("Distance between adjacent CpGs (bp)\n (Axis truncated at 300)") +
  ggtitle("Human reference genome (hg19)") +
  scale_y_continuous(labels = comma) +
  thesis_theme
g
ggsave("IPD_frequency_polygon_hg19.pdf", g, width = 16, height = 9)
```

# Many reads do not contain CpGs

I use the _Lister_ dataset to show that many reads from a whole-genome bisulfite-sequencing experiment do not contain any CpGs.

```{r}
sample_names <- c("ADS", "ADS-adipose", "ADS-iPSC", "FF", "FF-iPSC_19.11", 
                  "FF-iPSC_19.11+BMP4", "FF-iPSC_19.7", "FF-iPSC_6.9", 
                  "H1+BMP4", "H1_r1", "H1_r2", "H9", "H9_Laurent", "HSF1", 
                  "IMR90-iPSC", "IMR90_r1", "IMR90_r2")
y <- lapply(sample_names, function(sn) {
  fn <- paste0("../../Lister/", sn, "/1_tuples/hist/DM_", sn, 
               ".CG_per_read.hist")
  x <- read.table(fn, header = TRUE, as.is = TRUE)
  x[x$n == 0, "count"] / sum(as.numeric(x$count))
})
y <- data.frame("Sample" = sample_names,
                "Percentage" = round(100 * 
                                                   unlist(y, use.names = FALSE),
                                                 0)
                )
colnames(y) <- c("Sample", "Reads with no CpGs (%)")
write.table(y, "Lister_reads_with_no_CpGs.txt", quote = FALSE, 
            row.names = FALSE, col.names = TRUE, sep = "\t")
knitr::kable(y, align = "c")
```

# Session info

```{r}
library(devtools)
session_info()
```

