---
title: "Supplementary analyses supporting "Co-methylation Review""
author: "Peter Hickey"
date: "9/12/2014"
output: html_document
---

# Lyko et al. (2010)

Re-examaning the autocorrelations results of [http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1000506](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1000506).

```{r Lyko}
library(MethylationTuples)
library(ggplot2)
thesis_theme <- theme_bw(base_size = 20)

x <- readRDS("Lister/Lister_1_tuples_strand_collapsed.rds")
y <- x[seqnames(x) == "chr22", "FF"]
y <- y[!is.na(as.vector(methLevel(y)))]
beta <- rep(NA_real_, seqlengths(y)["chr22"])
beta[start(y)] <- methLevel(y)
beta_star <- rep(0, seqlengths(y)["chr22"])
beta_star[start(y)] <- methLevel(y)

system.time(beta_star_cor <- acf(beta_star, lag.max = 1000L, plot = FALSE))
system.time(beta_star_cor_ml <- 
              methLevelCor(y, pair_type = "all", ipd = seq_len(1000L)))

beta_star_cor_df <- data.frame(lag = beta_star_cor$lag, 
                               acf = beta_star_cor$acf)
g <- ggplot(aes(x = lag, y = acf), data = subset(beta_star_cor_df, lag > 0)) + 
  geom_line() + 
  thesis_theme + 
  ggtitle(expression(paste("FF (chr22): Autocorrelation of ", beta, "*"))) + 
  xlab("IPD") + 
  ylab("Pearson correlation")
ggsave("FF_chr21_acf_beta_star.pdf", g, height = 9, width = 16)

pdf(file = "FF_chr21_acf_beta_star_fourier.pdf", height = 9, width = 16)
spectrum(beta_star_cor$acf, main = expression(paste("FF (chr22): Fourier transform of autocorrelation of ", beta, "*")))
dev.off()
```

# Estimating the number of CpGs used by Lacey et. al (2013).

I want to estimate the number of CpGs used by \citet{Lacey:2013iy} to estimate the correlation of methylation levels. They used the chromosome 11 data from an RRBS experiment on a normal myotube cell line, _MTCTL2_. Unfortunately, there is no URL provided to these data. \citet{Lacey:2013iy} cite their data as having come from \citet{Tsumagari:2013gn}, which includes a sample "Normal Human Skeletal Muscle Myoblasts", _HSMM_ ([http://genome-mirror.duhs.duke.edu/ENCODE/cellTypes.html](http://genome-mirror.duhs.duke.edu/ENCODE/cellTypes.html), [http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/)). I believe this to be the same cell line. In any case, for my purposes I simply needed some comparable RRBS data, not necessarily the exact same sample as  _MTCTL2_.

There are approximately $60,000$ CpGs on chromosome 11 in these comparable RRBS data. The following R code shows how I computed this using three RRBS replicates of the _HSMM_ sample.

Firstly, I download the data and unzip it:
```{r, cache = FALSE, message = FALSE}
library(R.utils)
library(data.table)
dl_dir <- "../data/appendix/"
url <- 'http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/'
files <- paste0('wgEncodeHaibMethylRrbsHsmmtubefshdDukeSitesRep',
                1:3, '.bed.gz')
dests <- paste0(dl_dir, files)
x <- mapply(function(f, d) {
  if (!file.exists(gsub(".gz", "", paste0(dl_dir, f)))) {
    download.file(url = paste0(url, f), d)
    gunzip(filename = d, overwrite = TRUE)
    }
  fread(gsub(pattern = '.gz', replacement = '', x = d), sep = '\t',
        verbose = FALSE)
  }, files, dests, SIMPLIFY = FALSE)
```

Then I count the number of CpGs on chromosome 11 in each replicate:
```{r}
y <- rbindlist(x)
y[V1 == "chr11", .N, by = list(V4)]
```
